firmware/Core/Src/flash_bank/flash_bank.c

```diff
@@ -0,0 +1,124 @@
+ * If any part of the write operation fails, it breaks out of the loop and returns the error status.
+ * Finally, it locks the flash memory to disable the flash control register access.
+ *
+ * @param address The starting address in the flash memory where the data will be written.
+ * @param data A pointer to the data to be written to the flash memory.
+ * @param length The length of the data to be written.
+ *
+ * @return The status of the write operation. A value > 0 indicates an error, while a value of 0 indicates success.
+ */
+uint32_t Flash_Bank_Write(uint32_t address, uint8_t *data, uint32_t length)
+{
+    HAL_StatusTypeDef status;
+    uint32_t endAddress = address + length;
+
+    status = HAL_FLASH_Unlock();
+
+    if (status != HAL_OK)
+    {
+        return 1;
+    }
+
+    // Clear all FLASH flags before starting the operation
+    __HAL_FLASH_CLEAR_FLAG(FLASH_FLAG_ALL_ERRORS);
+
+    // Write data to flash
+    for (uint32_t currentAddress = address; currentAddress < endAddress; currentAddress += 8)
+    {
+        // currentAddress - address is the number of bytes already written
+        // const uint64_t double_word = *(uint64_t *)(data + (currentAddress - address));
+        uint8_t data_to_write[8] = {0};
+        memcpy(data_to_write, data + (currentAddress - address), 8);
+        const uint64_t double_word = *(uint64_t *)(data_to_write);
+        // Write the word to flash
+        status = HAL_FLASH_Program(FLASH_TYPEPROGRAM_DOUBLEWORD, currentAddress, double_word);
+        if (status != HAL_OK)
+        {
+            // If the program operation fails, break the loop
+            break;
+        }
+    }
+
+    // Lock the Flash to disable the flash control register access
+    HAL_FLASH_Lock();
+
+    return status;
+}
+
+/**
+ * Reads data from the flash memory.
+ *
+ * This function takes in an address, a pointer to a buffer to store the read data, and the length of the data to be read.
+ * It checks if the address is within the valid range of the flash memory and returns an error if it's not.
+ * If the address is valid, it reads the data from the flash memory and stores it in the provided buffer.
+ *
+ * @param address The starting address in the flash memory where the data will be read from.
+ * @param buffer A pointer to the buffer where the read data will be stored.
+ * @param length The length of the data to be read.
+ *
+ * @return 0 if the read operation is successful, 1 if the address is out of range.
+ */
+uint8_t Flash_Bank_Read(uint32_t address, uint8_t *buffer, uint32_t length)
+{
+    // Check if the address is within the valid range of the flash memory
+    if (address < Flash_Partitions_FLASH_BANK2 || (address + length) > (FLASH_BANK2_END))
+    {
+        return 1; // Return error if address is out of range
+    }
+
+    for (uint32_t i = 0; i < length - 1; i++)
+    {
+        buffer[i] = *(uint8_t *)(address + i);
+    }
+    buffer[length] = '\0';
```

Comment:
This doesn't make sense. Why is this array being null-terminated? Docstring indicates nothing of this behaviour, and it is non-idiomatic.

Reply:
this is a vestige of my testing, I mainly tested with characters so this made more sense then, but in reality, we would be passing in raw integer values, so I will remove it